### DISCUSSION TITLE
タッチパネルはどのように使用すればよいのでしょうか？

### DISCUSSION THREAD

**Discussion by kurokurokuro1** [lovyan03/LovyanGFX#539] [Q&A]
ESP32-S3と秋月電子販売のMSP2807でこちらのライブラリを使わせていただき苦戦しております。
描画についてはほぼ理解出来て目的とする数値や文字に図形を表示できるようになりました。
そこで今度はタッチパネルで描画した図形に対してスイッチの動作をさせたいと思い苦戦しております。

2_user_settingでタッチパネルの設定項目があり、使用するMSP2807はXPT2046で動く事はスケッチを書き込んで確認できました。
いざそれを自作スケッチに盛り込もうとした際、設定方法などが残念ながらこちらのgithubにもネット上にも見つけられず詰んでおります。

初期設定と座標に対してこのエリアにタッチがあったらフラグを立てるとかの方法をご教授頂ければ幸いです。
よろしくお願いいたします。

**Comment by embedded-kiddie** [comment]
Hi @kurokurokuro1 ,
Did you solve your question?
（こんにちは、この質問は解決しましたか？）

I recently discovered the greatness of LovyanGFX and started using it in my projects.
（最近LovyanGFXの素晴らしさを発見し、自分のプロジェクトに使い始めた者です。）

Since it looks like you have already tried `2_user_setting`, I think you have seen the calibration function `display.calibrateTouch()`.
（既に`2_user_setting`を試している様なので、キャリブレーション用関数`display.calibrateTouch()`は見ていると思います。）

You need to find out how this function handles the calibration results.
（この関数の中でキャリブレーション結果をどの様に扱っているのかを調べる必要があります。）

Once you search for this function, you will find that it is defined in [`LGFXBase.hpp`](https://github.com/lovyan03/LovyanGFX/blob/master/src/lgfx/v1/LGFXBase.hpp#L1459-L1472 "LovyanGFX/src/lgfx/v1/LGFXBase.hpp at master · lovyan03/LovyanGFX") and is written as follows:
（この関数を検索すると、`LGFXBase.hpp`に定義されていて、次のように記述されていることがわかります。）

```c++
    /// This requires a uint16_t array with 8 elements. ( or nullptr )
    template <typename T>
    void calibrateTouch(uint16_t *parameters, const T& color_fg, const T& color_bg, uint8_t size = 10)
    { // 第1引数 にuint16_t[8]のポインタを渡すことで、setTouchCalibrateで使用できるキャリブレーション値を得ることが出来る。;
      // この値をフラッシュ等に記録しておき、次回起動時にsetTouchCalibrateを使うことで、手作業によるキャリブレーションを省略できる。;
      calibrate_touch(parameters, _write_conv.convert(color_fg), _write_conv.convert(color_bg), size);
    }

    /// This requires a uint16_t array with 8 elements.
    /// calibrateTouchで得たキャリブレーション値を用いて設定を再現する。;
    void setTouchCalibrate(uint16_t *parameters)
    {
      panel()->setCalibrate(parameters);
    }
```

Then you can see that you need to do as the following comment says:
（すると次のコメントの通りにすれば良いことが分かります。）

> // By passing a pointer to uint16_t[8] as the first argument, you can obtain a calibration value that can be used with setTouchCalibrate. ;
> // By recording this value in a flash memory or other memory and using setTouchCalibrate the next time the device is started, manual calibration can be avoided. ;

> // 第1引数 にuint16_t[8]のポインタを渡すことで、setTouchCalibrateで使用できるキャリブレーション値を得ることが出来る。;
> // この値をフラッシュ等に記録しておき、次回起動時にsetTouchCalibrateを使うことで、手作業によるキャリブレーションを省略できる。;

That is to say, by passing a pointer to `uint16_t[8]` instead of `nullptr` as passed in the sample code of `2_user_setting`, you will get 8 calibration results in the array.
（つまり、`2_user_setting` のサンプルコードで渡された `nullptr` の代わりに `uint16_t[8]` へのポインタを渡せば、その配列を通して 8 つのキャリブレーション結果が取得できます。）

Then you can easily embed the results into your own sketch.
（後は、その結果を自身のスケッチに取り込むだけです。）

----------

You don't need to be fluent in English. Recently Google Translate has gotten better. So Please communicate in English in an open software repository that is available to the whole world.

If you had done that, I don't think this thread would have been left unattended for half a year.
Good luck!
